docs/se/guides/migration_3.x.adoc

```diff
@@ -0,0 +1,63 @@
+///////////////////////////////////////////////////////////////////////////////
+
+    Copyright (c) 2022 Oracle and/or its affiliates.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+///////////////////////////////////////////////////////////////////////////////
+
+= Helidon SE 3.x Upgrade Guide
+:description: Helidon SE 3.x Upgrade Guide
+:keywords: helidon, porting, migration, upgrade, incompatibilities
+:rootdir: {docdir}/../..
+
+include::{rootdir}/includes/se.adoc[]
+
+In Helidon 3 we have made some changes to APIs and runtime behavior. This guide
+will help you migrate a Helidon SE 2.x application to 3.x.
+
+include::{rootdir}/includes/guides/migration.adoc[]
+
+== New Routing
+
+Handling routes based on the protocol version is now possible by registering specific routes
+on routing builder.
+
+For further information check xref:../webserver.adoc[WebServer Documentation]
+
+== Http/2 Support
+
+Helidon now supports Http/2 upgrade from Http/1, Http/2 without prior knowledge
+and Http/2 with ALPN over TLS.
+
+For further information check xref:../webserver.adoc[WebServer Documentation]
+
+== WebSocket
+
+Helidon SE support is now based on the `WebSocketRouting` class which enables Helidon application to
+configure routing for both annotated and programmatic WebSocket endpoints.
+`TyrusSupport` is now deprecated.
+
+Use new dependency to enable `WebSocket`:
+
+[source,xml]
+----
+<dependency>
+    <groupId>io.helidon.webserver</groupId>
+    <artifactId>helidon-webserver-websocket</artifactId>
+</dependency>
+----
+
+== JEP-290 security hardening
+
+We are taking security seriously. Deserialization is now disabled by default. JEP-290 gives us tools to harden security when deserializing objects (using Java Serialization). In Helidon 3.x, we have decided to forbid any deserialization by default. Helidon does not use Java Serialization at all, and we consider it a high security risk. Java Serialization can still be used if required (requires additional setup to allow specific classes to be deserialized, or by explicitly switching off our JEP-290 support).
```

Comment:
We should link this once JEP-290 doc is added

Reply:
xref-ed